test/integration/consul-container/test/tproxy/tproxy_test.go

```diff
@@ -30,34 +33,74 @@ import (
 func TestTProxyService(t *testing.T) {
 	t.Parallel()
 
-	cluster, _, _ := topology.NewCluster(t, &topology.ClusterConfig{
-		NumServers:                1,
-		NumClients:                2,
-		ApplyDefaultProxySettings: true,
-		BuildOpts: &libcluster.BuildOptions{
-			Datacenter:             "dc1",
-			InjectAutoEncryption:   true,
-			InjectGossipEncryption: true,
-			// TODO(rb): fix the test to not need the service/envoy stack to use :8500
-			AllowHTTPAnyway: true,
-		},
-	})
+	cluster := createCluster(t, 2) // 2 client agent pods
 
 	clientService := createServices(t, cluster)
 	_, adminPort := clientService.GetAdminAddr()
 
 	libassert.AssertUpstreamEndpointStatus(t, adminPort, "static-server.default", "HEALTHY", 1)
 	libassert.AssertContainerState(t, clientService, "running")
-	assertHTTPRequestToVirtualAddress(t, clientService)
+	assertHTTPRequestToStaticServerVirtual(t, clientService)
 }
 
-func assertHTTPRequestToVirtualAddress(t *testing.T, clientService libservice.Service) {
-	timer := &retry.Timer{Timeout: 120 * time.Second, Wait: 500 * time.Millisecond}
+// TestTProxyPermissiveMTLS makes sure that a service in permissive mTLS mode accepts
+// non-mesh traffic on the upstream service's port.
+//
+// Steps:
+//   - Create a single server cluster
+//   - Create the static-server and static-client services in the mesh
+//   - Create an additional non-mesh container without a mesh service
+//   - In default/strict mTLS mode, make sure requests from static-client to
+//     static-server's virtual address succeed, but requests from the non-mesh
+//     service fail
+//   - In permissive mTLS mode, make sure requests from static-client to
+//     static-server's virtual addresss succeed, and requests from the non-mesh
+//     service to the static-server's normal address/port succeed.
+func TestTProxyPermissiveMTLS(t *testing.T) {
+	t.Parallel()
+
+	// Three client containers: static-client, static-server, and a non-mesh service.
+	// To simulate requests from outside the mesh, we use another client agent container
+	// without a mesh service, and run curl to simulate requests from outside the mesh.
+	cluster := createCluster(t, 3)
+
+	staticServerPod := cluster.Agents[1]
+	nonMeshPod := cluster.Agents[3]
+
+	clientService := createServices(t, cluster)
+	_, adminPort := clientService.GetAdminAddr()
+
+	libassert.AssertUpstreamEndpointStatus(t, adminPort, "static-server.default", "HEALTHY", 1)
+	libassert.AssertContainerState(t, clientService, "running")
+
+	// Validate mesh traffic to the virtual address succeeds in strict/default mTLS mode.
+	assertHTTPRequestToStaticServerVirtual(t, clientService)
```

Comment:
These names are a bit confusing to me at first glance. `assertHTTPRequestToStaticServerVirtual` is a static, virtual ip? Or a virtual IP that is not necessarily static for an otherwise static server? 

Is the `StaticServer` portion of this relevant or is the primary focus the `Virtual` address?

Reply:
Ah, `static-client` and `static-server` are the service names used in these integration tests. So this is a request to the Consul virtual hostname/IP of the "static-server" service.

I agree this is confusing out of context, so happy to change/rename (naming is hard)

Maybe it would make more sense if it was:

```go
assertHTTPRequestToVirtualAddress(t, clientService, "static-server")
```

Reply:
👍 This makes more sense to me at first glance. Definitely helps me follow the tests with less context.